src/ucp/tag/rndv.c

```diff
@@ -236,54 +308,64 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_proto_progress_rndv_get_zcopy, (self),
                  uct_pending_req_t *self)
 {
     ucp_request_t *rndv_req = ucs_container_of(self, ucp_request_t, send.uct);
+    ucp_ep_h ep             = rndv_req->send.ep;
     ucs_status_t status;
     size_t offset, length, ucp_mtu, align;
     const size_t max_iovcnt = 1;
     uct_iov_t iov[max_iovcnt];
     size_t iovcnt;
     ucp_rsc_index_t rsc_index;
     ucp_dt_state_t state;
+    ucp_lane_index_t lane;
+    ucp_lane_index_t lane_idx;
+    uct_rkey_t rkey;
+    int lanes;
 
-    if (ucp_ep_is_stub(rndv_req->send.ep)) {
+    if (ucp_ep_is_stub(ep)) {
         return UCS_ERR_NO_RESOURCE;
     }
 
-    if (!(ucp_tag_rndv_is_get_op_possible(rndv_req->send.ep, rndv_req->send.lane,
-                                          rndv_req->send.rndv_get.rkey_bundle.rkey))) {
+    if (!(ucp_tag_rndv_is_get_op_possible(ep, rndv_req))) {
         /* can't perform get_zcopy - switch to AM rndv */
-        if (rndv_req->send.rndv_get.rkey_bundle.rkey != UCT_INVALID_RKEY) {
-            uct_rkey_release(&rndv_req->send.rndv_get.rkey_bundle);
-        }
+        ucp_request_rndv_get_release(rndv_req);
         ucp_rndv_recv_am(rndv_req, rndv_req->send.rndv_get.rreq,
                          rndv_req->send.rndv_get.remote_request,
                          rndv_req->send.length);
 
         return UCS_INPROGRESS;
     }
 
-    rsc_index = ucp_ep_get_rsc_index(rndv_req->send.ep, rndv_req->send.lane);
-    align     = rndv_req->send.ep->worker->ifaces[rsc_index].attr.cap.get.opt_zcopy_align;
-    ucp_mtu   = rndv_req->send.ep->worker->ifaces[rsc_index].attr.cap.get.align_mtu;
-
-    ucs_trace_data("ep: %p try to progress get_zcopy for rndv get. rndv_req: %p. lane: %d",
-                   rndv_req->send.ep, rndv_req, rndv_req->send.lane);
-
     /* rndv_req is the internal request to perform the get operation */
-    if (rndv_req->send.state.dt.dt.contig.memh == UCT_MEM_HANDLE_NULL) {
+    if (!ucp_dt_have_rndv_lanes(&rndv_req->send.state.dt)) {
         /* TODO Not all UCTs need registration on the recv side */
-        UCS_PROFILE_REQUEST_EVENT(rndv_req->send.rndv_get.rreq, "rndv_recv_reg", 0);
-        status = ucp_request_send_buffer_reg(rndv_req, rndv_req->send.lane);
-        ucs_assert_always(status == UCS_OK);
+        ucp_request_rndv_reg(rndv_req);
+        rndv_req->send.reg_rsc = UCP_NULL_RESOURCE;
```

Comment:
this line should be moved inside ucp_request_rndv_reg() , and also removed from ucp_request_rndv_dereg()